### ISSUE TITLE
2016/dataset.week.3/stage.2.blastnfilter failed

### ISSUE THREAD

**Issue by coleslaw481** [drugdata/D3R#5]
Without any error message 2016/dataset.week.3/stage.2.blastnfilter failed with non zero exit code. (Running version 0.8.6)  No error message was seen anywhere.
Attached is a zip file of the 2nd run that failed. (it was run twice with same effect)

ls -lrt of directory:

$ ls -lrt
total 11740
-rw-r--r-- 1 d3r d3r     615 Jan 19 09:14 summary.txt
-rw-r--r-- 1 d3r d3r       0 Jan 19 09:14 start
-rw-r--r-- 1 d3r d3r     308 Jan 19 09:20 5c1a.txt
-rw-r--r-- 1 d3r d3r     111 Jan 19 09:30 5hav.txt
-rw-r--r-- 1 d3r d3r     226 Jan 19 09:33 5fpv.txt
-rw-r--r-- 1 d3r d3r     317 Jan 19 09:35 5fpu.txt
-rw-r--r-- 1 d3r d3r     184 Jan 19 09:36 5fn0.txt
-rw-r--r-- 1 d3r d3r     311 Jan 19 09:38 5fmq.txt
-rw-r--r-- 1 d3r d3r     328 Jan 19 09:39 5fmm.txt
-rw-r--r-- 1 d3r d3r     221 Jan 19 09:48 5fg4.txt
-rw-r--r-- 1 d3r d3r     335 Jan 19 10:04 5fet.txt
-rw-r--r-- 1 d3r d3r     229 Jan 19 10:08 5fea.txt
-rw-r--r-- 1 d3r d3r     191 Jan 19 10:16 5fe9.txt
-rw-r--r-- 1 d3r d3r     254 Jan 19 10:27 5fe8.txt
-rw-r--r-- 1 d3r d3r     198 Jan 19 10:39 5fe7.txt
-rw-r--r-- 1 d3r d3r     169 Jan 19 10:50 5fe6.txt
-rw-r--r-- 1 d3r d3r     171 Jan 19 11:02 5fe5.txt
-rw-r--r-- 1 d3r d3r     179 Jan 19 11:13 5fe4.txt
-rw-r--r-- 1 d3r d3r     174 Jan 19 11:25 5fe3.txt
-rw-r--r-- 1 d3r d3r     259 Jan 19 11:36 5fe2.txt
-rw-r--r-- 1 d3r d3r     164 Jan 19 11:48 5fe1.txt
-rw-r--r-- 1 d3r d3r     199 Jan 19 11:59 5fe0.txt
-rw-r--r-- 1 d3r d3r     254 Jan 19 12:11 5fdz.txt
-rw-r--r-- 1 d3r d3r     274 Jan 19 12:11 5fbs.txt
-rw-r--r-- 1 d3r d3r     400 Jan 19 12:19 5f20.txt
-rw-r--r-- 1 d3r d3r     368 Jan 19 12:30 5f1z.txt
-rw-r--r-- 1 d3r d3r     286 Jan 19 12:32 5ext.txt
-rw-r--r-- 1 d3r d3r     308 Jan 19 12:34 5exs.txt
-rw-r--r-- 1 d3r d3r     528 Jan 20 00:31 5etx.txt
-rw-r--r-- 1 d3r d3r     141 Jan 20 00:39 5es3.txt
-rw-r--r-- 1 d3r d3r     509 Jan 20 12:33 5eqq.txt
-rw-r--r-- 1 d3r d3r 2637015 Jan 20 12:48 blastnfilter.log
-rw-r--r-- 1 d3r d3r     356 Jan 20 12:48 5eqb.txt
-rw-r--r-- 1 d3r d3r    3487 Jan 20 13:48 blastnfilter.py.stdout
-rw-r--r-- 1 d3r d3r  885936 Jan 20 13:48 blastnfilter.py.stderr
-rw-r--r-- 1 d3r d3r       0 Jan 20 13:49 postanalysis.py.stdout
-rw-r--r-- 1 d3r d3r    2657 Jan 20 13:49 postanalysis.py.stderr
-rw-r--r-- 1 d3r d3r       0 Jan 20 13:49 error
-rw-r--r-- 1 d3r d3r 8336944 Jan 20 13:49 dockable.xlsx

[stage.2.blastnfilter.zip](https://github.com/nbcrrolls/D3R/files/101616/stage.2.blastnfilter.zip)

**Comment by coleslaw481** [comment]
Another week with memory failure.

blastnfilter failed this week (/data/celpp/2016/dataset.week.23/stage.3.blastnfilter) with an out of memory error (according to last lines of output in blastnfilter.py.stderr file)

**Comment by coleslaw481** [comment]
/celpp/data/2017/dataset.week.26/stage.3.blastnfilter also failed with out of memory error:

Output from **blastnfilter.py.stderr**:
```Bash
2017-06-24 22:09:47,003 DEBUG d3r.filter.filter In filter_for_highest_tanimoto()
2017-06-24 22:09:47,004 DEBUG d3r.utilities.run set_query
2017-06-24 22:09:48,032 DEBUG d3r.utilities.run Blasting query:  5gmp
2017-06-24 22:09:48,081 DEBUG d3r.utilities.run query.run_blast
2017-06-24 22:09:48,083 DEBUG d3r.blast.query In run_blast()
2017-06-24 22:09:48,085 DEBUG d3r.blast.query Running blast_monomer
2017-06-24 22:09:48,087 DEBUG d3r.blast.query There are 1 to analyze but this code will only analyze the first sequence???
2017-06-24 22:09:48,088 DEBUG d3r.blast.query Writing fasta file: /celpp/data/2017/dataset.week.26/stage.3.blastnfilter/5gmp_5gmp_1.fasta
2017-06-24 22:09:48,137 DEBUG d3r.blast.query Starting blastp of /celpp/data/2017/dataset.week.26/stage.3.blastnfilter/5gmp_5gmp_1.fasta
2017-06-24 22:09:48,138 DEBUG d3r.blast.query Running blastp /celpp/data/2017/dataset.week.26/stage.3.blastnfilter/5gmp_5gmp_1.fasta
2017-06-24 22:09:48,156 ERROR d3r.blastnfilter Error caught exception
Traceback (most recent call last):
  File "/usr/bin/blastnfilter.py", line 69, in main
    run.run(options)
  File "/usr/lib/python2.6/site-packages/d3r/utilities/run.py", line 150, in run
    query = blast_the_query(query, pdb_db, pdb_path, fasta, out_dir, compinchi)
  File "/usr/lib/python2.6/site-packages/d3r/utilities/run.py", line 57, in blast_the_query
    records = query.run_blast(pdb_db, out_dir)
  File "/usr/lib/python2.6/site-packages/d3r/blast/query.py", line 114, in run_blast
    record = self.blast_monomer(pdb_db, out_dir)
  File "/usr/lib/python2.6/site-packages/d3r/blast/query.py", line 141, in blast_monomer
    record = self.blastp(fasta, pdb_db)
  File "/usr/lib/python2.6/site-packages/d3r/blast/query.py", line 217, in blastp
    std_out, std_err = cline()
  File "/usr/lib64/python2.6/site-packages/Bio/Application/__init__.py", line 495, in __call__
    shell=use_shell)
  File "/usr/lib64/python2.6/subprocess.py", line 642, in __init__
    errread, errwrite)
  File "/usr/lib64/python2.6/subprocess.py", line 1144, in _execute_child
    self.pid = os.fork()
OSError: [Errno 12] Cannot allocate memory
```

**Comment by j-wags** [comment]
@coleslaw481 , I'm having a lot of trouble figuring this out. Could we meet this week, like Wednesday?

**Comment by coleslaw481** [comment]
Sure. Anytime before 2pm is fine with me.

**Comment by coleslaw481** [comment]
Band-aid fix made in commit: 12d79ac07f30d0977f05480d8ff5e62253965723
Basically using fork in d3r/utilities/run.py to run the main blastnfilter code in run() function in a separate child process for each query. This seems to eliminate problem with running out of memory, but its not very elegant.
